src/write.c

```diff
@@ -3090,6 +3090,18 @@ static avifResult avifRWStreamWriteProperties(avifItemPropertyDedup * const dedu
             // We don't add an 'lsel' property since many decoders do not support it and will reject the image,
             // see https://github.com/AOMediaCodec/libavif/pull/2429
         }
+        for (size_t i = 0; i < itemMetadata->numProperties; i++) {
+            avifItemPropertyDedupStart(dedup);
+            avifImageItemProperty * prop = &(itemMetadata->properties[i]);
+            avifBoxMarker propMarker;
+            AVIF_CHECKRES(avifRWStreamWriteBox(&dedup->s, (const char *)prop->boxtype, AVIF_BOX_SIZE_TBD, &propMarker));
```

Comment:
I wonder if we should check for custom box types (for example forbid `meta`). Otherwise it would be possible to generate invalid files with the libavif API.

With or without this check, we should fuzz this new API. I think a new dedicated fuzz target would be easy. It needs to check that encoding any API-compliant custom prop garbage is a success, and that decoding the encoded output is a success when expected (that is the hard part to determine) or a graceful error.  
Maybe an approach similar to `ArbitraryAvifImageWithGainMap()` would be a good fit.

Reply:
I added a note to the API documentation about not adding known properties.

The other parts are still to be done.

Reply:
> I added a note to the API documentation about not adding known properties.

"Known properties" is likely too vague as a criterion.

I would still prefer the encoder to check for validity. I can think of ways:
- Check that there is no forbidden box type (an approximate list should be rather quick to gather, for example [here](https://aomediacodec.github.io/av1-avif/#avif-boxes) up to level 2)
- Check that the box type makes sense (`[a-zA-Z0-9! ]{4}` or something)
- Check that the uuid makes sense (is that possible?)
- Check that there is no property with the same box type assigned twice to the same item (assuming nobody needs that besides `colr` which is set by libavif)
- Call `avifParse()` within `avifEncode()` when custom properties are set (a bit too much of a change maybe)

@wantehchang What do you think?

> The other parts are still to be done.

You meant in this PR or in another one?

Reply:
> > The other parts are still to be done.
> 
> You meant in this PR or in another one?

I mean the other parts of this comment. 

I'm doubtful about the value of checking for known boxes on encode, but will try it. My concern is that we're trying to protect programmers from themselves. They can always make junk files (e.g. by changing the results of the encode before writing it out) and we're potentially making the encode slower for  limited protection.

I would like clarification on why you only wanted up to level 2 though - if we're going to check properties, why not check for adding known properties?  Possibly that would be caught by the fourth point (no duplicates), but I'm -0 on that, because there are likely uses for that (e.g. `udes` with different languages is supported in 23008-12, and maybe similar use of that pattern in other documents).

There are pretty limited options for checking a UUID, but I could see how much hassle that is.

Instead of requiring `avifParse()`, we could suggest that as an option for users. Again, trying to limit the impact on encode performance.

Reply:
I've added argument checks and associated test code. The checks are at set time, so the failure result is closer to the problem.

I still need to add documentation about using `avifParse`.

Reply:
> With or without this check, we should fuzz this new API. I think a new dedicated fuzz target would be easy. It needs to check that encoding any API-compliant custom prop garbage is a success, and that decoding the encoded output is a success when expected (that is the hard part to determine) or a graceful error.
> Maybe an approach similar to `ArbitraryAvifImageWithGainMap()` would be a good fit.

I can't say it was easy. However I have added the fuzzer.